org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/MavenBuildSupport.java

```diff
@@ -123,4 +137,68 @@ public boolean fileChanged(IResource resource, CHANGE_TYPE changeType, IProgress
 		}
 		return IBuildSupport.super.fileChanged(resource, changeType, monitor) || isBuildFile(resource);
 	}
+
+	/* (non-Javadoc)
+	 * @see org.eclipse.jdt.ls.core.internal.managers.IBuildSupport#getSource(org.eclipse.jdt.core.IClassFile, org.eclipse.core.runtime.IProgressMonitor)
+	 */
+	@Override
+	public void getSource(IClassFile classFile, IProgressMonitor monitor) throws CoreException {
+		if (classFile == null) {
+			return;
+		}
+		IJavaElement element = classFile;
+		while (element.getParent() != null) {
+			element = element.getParent();
+			if (element instanceof IPackageFragmentRoot) {
+				final IPackageFragmentRoot fragment = (IPackageFragmentRoot) element;
+				IPath attachmentPath = fragment.getSourceAttachmentPath();
+				if (attachmentPath != null && !attachmentPath.isEmpty() && attachmentPath.toFile().exists()) {
+					break;
+				}
+				if (fragment.isArchive()) {
+					IProject project = fragment.getJavaProject().getProject();
+					Boolean downloaded = cache.getIfPresent(project.getName());
+					if (downloaded == null) {
+						IClasspathManager buildpathManager = MavenJdtPlugin.getDefault().getBuildpathManager();
+						buildpathManager.scheduleDownload(fragment, true, true);
+						waitForDownload(monitor);
+						buildpathManager.scheduleDownload(project, true, true);
```

Comment:
no need to expand the downloads to the whole projects, we're trying to keep things lazy here